packages/react-aria-components/src/Collection.tsx

```diff
@@ -905,26 +891,61 @@ export interface SectionProps<T> extends Omit<SharedSectionProps<T>, 'children'
   dependencies?: any[]
 }
 
-function Section<T extends object>(props: SectionProps<T>, ref: ForwardedRef<HTMLElement>): JSX.Element | null {
-  let children = useCollectionChildren(props);
-  return useSSRCollectionNode('section', props, ref, null, children);
+interface SectionContextValue {
+  render: (section: Node<any>, props: SectionProps<any>, ref: ForwardedRef<HTMLElement>) => ReactElement
 }
 
-const _Section = /*#__PURE__*/ (forwardRef as forwardRefType)(Section);
-export {_Section as Section};
+export const SectionContext = createContext<SectionContextValue | null>(null);
+
+export const Section = /*#__PURE__*/ createBranchComponent('section', <T extends object>(props: SectionProps<T>, ref: ForwardedRef<HTMLElement>, section: Node<T>): JSX.Element => {
+  let {render} = useContext(SectionContext)!;
+  return render(section, props, ref);
+});
 
 export const CollectionContext = createContext<CachedChildrenOptions<unknown> | null>(null);
-export const CollectionRendererContext = createContext<CollectionProps<any>['children']>(null);
 
 /** A Collection renders a list of items, automatically managing caching and keys. */
 export function Collection<T extends object>(props: CollectionProps<T>): JSX.Element {
   let ctx = useContext(CollectionContext)!;
   props = mergeProps(ctx, props);
   props.dependencies = (ctx?.dependencies || []).concat(props.dependencies);
-  let renderer = typeof props.children === 'function' ? props.children : null;
-  return (
-    <CollectionRendererContext.Provider value={renderer}>
-      {useCollectionChildren(props)}
-    </CollectionRendererContext.Provider>
-  );
+  return <>{useCollectionChildren(props)}</>;
+}
+
+export function createLeafComponent<T extends object, P extends object, E extends Element>(type: string, render: (props: P, ref: ForwardedRef<E>, node: Node<T>) => JSX.Element) {
```

Comment:
You mentioned you weren't happy about the api here for these, did you have any other specific concerns other than possibly combining them more? To me, using these for custom collections already feels like a pretty advanced use case that needs a deeper understanding of how the RAC collections work so not sure how to simplify them more tbh

Reply:
If its in any way useful I would give my two cents here, after studying this file for the past couple days and re-building this whole thing as well as a test suite to all of this locally.

In my experience the naming changes introduced in this PR have made the API much more approachable compared to the prior direct exposure of `useSSRCollectionNode()` in both leaf and branch components. The concept of a tree and linked list is likely much more familiar to most software engineers than the `NodeList` and `Document` interface. 

I definitely agree with LFDanLu however that a deep understanding of `Collection.tsx` is required to build any sort of custom implementation.

As dumb as it may sound, my biggest learning obstructor about the current API has been that literally everything is called a `Node` of some sort. Whether it be a `BaseNode` to traverse the fake document, an `ElementNode` managing the lifecycle of a `NodeValue` implementing react-stately's `Node` interface or be it one of the many hooks, class methods or properties involved in the implementation itself. Everything sounded and felt very indifferent to me.

It was one of the first things i adjusted in order to get a better grasp of what's going on from a higher level, after which it was rather smooth.

Reply:
Thanks for the input, we'd still like to eliminate the useShallowComponent since that name doesn't make a ton of sense. We'd discussed something like this in order to make a meaningfully named option, though haven't quite gotten the types worked out for it.
```
export function createLeafComponent<T extends object, P extends object, E extends Element>(type: string, render: (props: P, ref: ForwardedRef<E>, node: Node<T>) => JSX.Element, allowsOutsideCollection?: false);
export function createLeafComponent<T extends object, P extends object, E extends Element>(type: string, render: (props: P, ref: ForwardedRef<E>, node?: Node<T>) => JSX.Element, allowsOutsideCollection: true);
export function createLeafComponent<T extends object, P extends object, E extends Element>(type: string, render: (props: P, ref: ForwardedRef<E>, node?: Node<T>) => JSX.Element, allowsOutsideCollection: boolean = false) {
```

Reply:
Got it, I understand the desire to remove `createShallowComponent()` 👍.
 
Personally i am not a big fan of hiding things behind boolean flags, since they are hard to read and discover, but I also don’t have a better suggestion in mind at current.

PS: I must have accidentally deleted my previous comment, so excuse the messed up timeline.

Reply:
Yep, we're not totally enamored with it, either. But we think it's better than `createShallowComponent`, it's descriptive and shows it's only allowed for a leaf. If anyone has a better idea, we're all ears.

No worries about the deleted comment.